CRM/Utils/SQL/TempTable.php

```diff
@@ -128,13 +128,25 @@ public function getName() {
    * @return CRM_Utils_SQL_TempTable
    */
   public function createWithQuery($selectQuery) {
-    $sql = sprintf('%s %s %s AS %s',
-      $this->toSQL('CREATE'),
-      $this->memory ? self::MEMORY : self::INNODB,
-      $this->getUtf8String(),
-      ($selectQuery instanceof CRM_Utils_SQL_Select ? $selectQuery->toSQL() : $selectQuery)
-    );
-    CRM_Core_DAO::executeQuery($sql, [], TRUE, NULL, TRUE, FALSE);
+    try {
+      $sql = sprintf(
+        '%s %s %s AS %s',
+        $this->toSQL('CREATE'),
+        $this->memory = self::MEMORY,
+        $this->getUtf8String(),
+        ($selectQuery instanceof CRM_Utils_SQL_Select ? $selectQuery->toSQL() : $selectQuery)
+      );
+      CRM_Core_DAO::executeQuery($sql, [], TRUE, NULL, TRUE, FALSE);
+    } catch (Exception $e) {
```

Comment:
Thanks for the PR. The style police are saying there needs to be a newline after the closing `}` and before the `catch`.

The other error is that changing `$this->memory ? self::MEMORY : self::INNODB,` to always be `$this->memory = self::MEMORY,` causes some problems with temporary tables that were already specified as innodb. It needs to observe that preference.

Reply:
You can see the errors if you click on the details links on the red test lines that appear at the bottom of the conversation tab.

Reply:
If y'all want to decide upfront (e.g. decide when `TempTable::__construct()` initializes the value of `$this->memory`), then might find a helper like this useful: https://gist.github.com/totten/48e2feddbd4af38a3a580821e47f90ee

Reply:
Thanks @totten, I wasn't aware of that helper function. However, I already tried to check upfront with a query to show engines and the problem is that DO clusters do show MEMORY as available engines but in fact it is not available: they set disabled_storage_engines mysql setting with "ARCHIVE,BLACKHOLE,CSV,MEMORY,MRG_MYISAM,MyISAM". There is no way to use MEMORY engine even by setting internal_tmp_mem_storage_engine to MEMORY.

This is the result of SHOW ENGINES in my DO cluster, which triggers an error when using MEMORY engine:

![dbaas-codinesparticipa___codinesparticipa_-_MySQL_8_0_30](https://github.com/civicrm/civicrm-core/assets/105581/7d12722b-2c11-48a0-b59c-f3fb8479439f)

Reply:
Just confirmed that it doesn't work wit something like this:

```
public function createWithColumns($columns) {
    $this->memory = $this->supportsEngine('MEMORY') ? self::MEMORY : self::INNODB;
    $sql = sprintf('%s (%s) %s %s',
      $this->toSQL('CREATE'),
      $columns,
      $this->memory,
      $this->getUtf8String()
    );
    CRM_Core_DAO::executeQuery($sql, [], TRUE, NULL, TRUE, FALSE);
    $this->createSql = $sql;
    return $this;
  }
```
and adding that helper function within the same file.

Reply:
@robertgarrigos Ooh, wow. So it's reported as available - but doesn't actually work. I can see how you got the `try`/`catch`  then.

Reply:
Is there any update here?

Reply:
Memory engine can be reported as available, but it can be disabled by the db cloud admin.
```
mysql> SELECT @@disabled_storage_engines;
+------------------------------------------------+
| @@disabled_storage_engines                     |
+------------------------------------------------+
| ARCHIVE,BLACKHOLE,CSV,MEMORY,MRG_MYISAM,MyISAM |
+------------------------------------------------+
```